Port `SslHandler` from `ByteBuf` to `Buffer`

chrisvest opened the pull request:
Motivation:
We are replacing `ByteBuf` with `Buffer` in Netty.

Modification:
`SslHandler` has been changed to use `Buffer` instead of `ByteBuf`.
This in turn caused many other changes that were required to make this transition.

Result:
`SslHandler` now uses `Buffer` instead of `ByteBuf`.

This PR is a draft because there are many tests that are not passing yet, with this change.

chrisvest commented:
Also, there are probably a good number of changes in here that we want to extract into separate PRs.

hyperxpro commented:
Mission accomplished? @chrisvest :D

violetagg commented on testsuite/src/main/java/io/netty5/testsuite/transport/socket/SocketStartTlsTest.java:
```diff
@@ -130,11 +131,13 @@ public static void testStartTlsNotAutoRead(ServerBootstrap sb, Bootstrap cb,
     private static void testStartTls(ServerBootstrap sb, Bootstrap cb,
                                      SslContext serverCtx, SslContext clientCtx, boolean autoRead) throws Throwable {
+        sb.childOption(ChannelOption.RCVBUF_ALLOCATOR_USE_BUFFER, true);
+        cb.option(ChannelOption.RCVBUF_ALLOCATOR_USE_BUFFER, true);
```
Can we reuse `io.netty5.testsuite.transport.AbstractComboTestsuiteTest#enableNewBufferAPI`?

chrisvest commented:
Some of the changes here have been extracted into #12248, so I want to rebase this on that before merging.

chrisvest commented on handler/src/main/java/io/netty5/handler/ssl/PemPrivateKey.java:
```diff
@@ -132,62 +127,32 @@ public boolean isSensitive() {
-        SslUtils.zerooutAndRelease(content);
+        // before we're releasing the underlying Buffer
+        // TODO cannot do this when the buffer is read-only
+//        SslUtils.zeroout(getBuffer());
```
I'm working on a follow-up PR to address this

chrisvest replied:
#12255 will make this possible to fix

chrisvest commented on handler/src/main/java/io/netty5/handler/ssl/PemValue.java:
```diff
@@ -46,61 +43,31 @@ public boolean isSensitive() {
+    public void close() {
+        if (sensitive) {
+            // TODO cannot do this when the buffer is read-only
+//            SslUtils.zeroout(getBuffer());
```
I'm working on a follow-up PR to address this

chrisvest replied:
#12255 will make this possible to fix

normanmaurer commented on buffer/src/main/java/io/netty5/buffer/ByteBufUtil.java:
```diff
@@ -217,7 +200,7 @@ public static void copy(AsciiString src, int srcIdx, ByteBuf dst, int length) {
                 }
                 BYTEPADDING[i] = buf.toString();
-            }
+        }
```
nit: seems like formatting is off. 
```suggestion
            }
```

normanmaurer commented on codec-http2/src/test/java/io/netty5/handler/codec/http2/Http2MultiplexTransportTest.java:
```diff
@@ -371,7 +374,7 @@ protected void initChannel(Channel ch) {
                 @Override
                 protected void initChannel(Channel ch) {
-                    ch.pipeline().addLast(clientCtx.newHandler(ch.alloc()));
+                    ch.pipeline().addLast(clientCtx.newHandler(ch.bufferAllocator()));
```
Why we don't need the `BufferConversionHandler.bufferToByteBuf()` here but we need it above ?

chrisvest replied:
Good question! Turns out in the case above, the `SslHandler` receives a `write()` call _before_ `channelActive` is called, which is the decision to implicitly convert buffers or not is made. Perhaps this implicit conversion should either always happen, or never happen. Dynamic detection doesn't look so reliable.

chrisvest replied:
Let's remove the implicit conversion in a separate PR.

chrisvest replied:
PR: https://github.com/netty/netty/pull/12297

normanmaurer commented on handler/src/main/java/io/netty5/handler/ssl/ConscryptAlpnSslEngine.java:
```diff
@@ -41,21 +44,22 @@
     }
 
-    static ConscryptAlpnSslEngine newServerEngine(SSLEngine engine, ByteBufAllocator alloc,
+    static ConscryptAlpnSslEngine newServerEngine(SSLEngine engine, io.netty5.buffer.api.BufferAllocator alloc,
```
nit: import `BufferAllocator` ?

chrisvest replied:
Conscrypt has their own thing called `BufferAllocator`, so one of them needs to be qualified.

normanmaurer commented on handler/src/main/java/io/netty5/handler/ssl/EngineWrapper.java:
```diff
@@ -0,0 +1,222 @@
+        if (in == null || in.readableBytes() == 0) {
+            inputs = singleEmptyBuffer;
+        } else {
+            if (in.isDirect() == useDirectBuffer) {
```
nit: you can make this an `} else if (....) {`

normanmaurer commented on handler/src/main/java/io/netty5/handler/ssl/EngineWrapper.java:
```diff
@@ -0,0 +1,222 @@
+        if (out == null || out.writableBytes() == 0) {
+            outputs = singleEmptyBuffer;
+        } else {
+            if (out.isDirect() == useDirectBuffer) {
```
nit: you can make this an `else if`

normanmaurer commented on handler/src/main/java/io/netty5/handler/ssl/EngineWrapper.java:
```diff
@@ -0,0 +1,222 @@
+        if (result != null) { // Result can be null if the operation failed.
+            if (in != null) {
+                in.skipReadable(result.bytesConsumed());
+                Reference.reachabilityFence(in);
```
please add comments why we use a `reachableFence(...)`. The same comment holds true for below.

chrisvest replied:
Yes, this also needs to be restructured a bit.

normanmaurer commented on handler/src/main/java/io/netty5/handler/ssl/EngineWrapper.java:
```diff
@@ -0,0 +1,222 @@
+                if (writeBack) {
+                    assert outputs.length == 1;
+                    ByteBuffer buf = outputs[0];
+                    while (buf.remaining() > Long.BYTES) {
```
```suggestion
                    while (buf.remaining() >= Long.BYTES) {
```

normanmaurer commented on handler/src/main/java/io/netty5/handler/ssl/EngineWrapper.java:
```diff
@@ -0,0 +1,222 @@
+                    while (buf.remaining() > Long.BYTES) {
+                        out.writeLong(buf.getLong());
+                    }
+                    while (buf.hasRemaining()) {
```
You can also do the same as above but with `Integer.BYTES`

normanmaurer commented on handler/src/main/java/io/netty5/handler/ssl/EngineWrapper.java:
```diff
@@ -0,0 +1,222 @@
+
+    @Override
+    public boolean process(int index, ReadableComponent component) {
+        // Some SSLEngine implementations require their input buffers be mutable. Let's try to accommodate.
```
Let's add a link to the openjdk bug .

normanmaurer commented on handler/src/main/java/io/netty5/handler/ssl/OptionalSslHandler.java:
```diff
@@ -86,15 +85,15 @@ protected String newSslHandlerName() {
      * Override to configure the SslHandler eg. {@link SSLParameters#setEndpointIdentificationAlgorithm(String)}.
      * The hostname and port is not known by this method so servers may want to override this method and use the
-     * {@link SslContext#newHandler(ByteBufAllocator, String, int)} variant.
+     * {@link SslContext#newHandler(io.netty5.buffer.api.BufferAllocator, String, int)} variant.
```
import this

normanmaurer commented on handler/src/main/java/io/netty5/handler/ssl/PemValue.java:
```diff
@@ -46,61 +43,31 @@ public boolean isSensitive() {
+    public Buffer content() {
+        if (!isAccessible()) {
+            throw new IllegalStateException("PemValue is closed.");
         }
```
Why not do this in `getBuffer()` ?

chrisvest replied:
In a separate PR.

normanmaurer commented on handler/src/main/java/io/netty5/handler/ssl/ReferenceCountedOpenSslEngine.java:
```diff
@@ -507,8 +517,8 @@ public final synchronized void shutdown() {
      * Calling this function with src.remaining == 0 is undefined.
      */
     private int writePlaintextData(final ByteBuffer src, int len) {
+        assert len > 0 && src.remaining() > 0;
```
use two assert statements so its easier to see what was the cause of it ?

normanmaurer commented on handler/src/main/java/io/netty5/handler/ssl/SslClientHelloHandler.java:
```diff
@@ -171,21 +178,19 @@ private void releaseHandshakeBuffer() {
-        if (buffer != null) {
-            buffer.release();
-        }
+    private static void releaseIfNotNull(Buffer buffer) {
```
nit: you could just directly call `dispose(...)` and remove this method

normanmaurer commented on handler/src/main/java/io/netty5/handler/ssl/VectoredUnwrap.java:
```diff
@@ -0,0 +1,38 @@
+ * Optional extra interface that {@link javax.net.ssl.SSLEngine} sub-classes can implement, if they support a faster
+ * way of unwrapping multiple input buffers into multiple output buffers.
+ */
+public interface VectoredUnwrap {
```
Do we want to keep this package private for now ?

normanmaurer commented on handler/src/test/java/io/netty5/handler/ssl/SslUtilsTest.java:
```diff
@@ -17,45 +17,42 @@
+                ByteBuffer[] byteBuffers = { component.readableBuffer() };
+                assertEquals(getEncryptedPacketLength(byteBuffers, 0),
+                             getEncryptedPacketLength(copy, 0));
+                return false;
```
why false ?

chrisvest replied:
There should only be one component. We can make that an assert, though, so it doesn't sneak past us if it ever changes.

normanmaurer commented on microbench/src/main/java/io/netty5/microbench/handler/ssl/AbstractSslEngineBenchmark.java:
```diff
@@ -15,7 +15,7 @@
 package io.netty5.microbench.handler.ssl;
 
-import io.netty.buffer.ByteBufAllocator;
+import io.netty5.buffer.api.BufferAllocator;
```
This would be actually interesting... Can you run the benchmark for 4.1 and compare to this branch ?

chrisvest replied:
Performance is currently half of 4.1, because the `PooledBufferAllocator` has gotten slower, it seems.

normanmaurer reviewed (changes requested):
Great work! Just a few suggestions / comments / questions

franz1981 commented:
FYI now playing with `SslHandlerEchoBenchmark` :)

franz1981 commented:
Please @chrisvest can you share your JMH parameters so I can run the same benchmark as you?
(For both netty 4.1 and netty 5 versions)?

For netty 5 I'm using the IDEA JMH plugin and running it with
```
io.netty5.microbench.handler.ssl.SslHandlerEchoBenchmark.* -p bufferType=HEAP -p cipher=TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 -p messageSize=128 -p numWrites=10 -p sslProvider=JDK -prof "async:output=flamegraph;dir=/tmp;libPath=/home/forked_franz/IdeaProjects/async-profiler/build/libasyncProfiler.so" -prof perfasm --jvmArgs="-server -dsa -da -ea:io.netty5... -XX:+HeapDumpOnOutOfMemoryError -XX:+UnlockDiagnosticVMOptions -XX:+DebugNonSafepoints -Dio.netty5.leakDetection.level=disabled -Dio.netty5.buffer.leakDetectionEnabled=false -Dio.netty5.buffer.lifecycleTracingEnabled=false -Xms768m -Xmx768m -XX:MaxDirectMemorySize=768m -XX:BiasedLockingStartupDelay=0 -Djmh.executor=CUSTOM -Djmh.executor.class=io.netty5.microbench.util.AbstractMicrobenchmark$HarnessExecutor"
```
with `# VM version: JDK 11.0.14.1, OpenJDK 64-Bit Server VM, 11.0.14.1+1`

and I've noticed that it uses `ByteBufferMemoryManager` ...is it correct?

franz1981 commented:
sent https://github.com/netty/netty/pull/12276 to fix https://github.com/netty/netty/pull/12221#issuecomment-1087513252

franz1981 reviewed (changes requested):
Suggest to bring in https://github.com/netty/netty/pull/12275 and https://github.com/netty/netty/pull/12276 in order to check what's the right numbers looks like

chrisvest commented:
Ran `SslHandlerEchoBenchmark` with JDK 11.0.7 and #12275, #12276, #12277, and this is what I got:

```
Benchmark                           (bufferType)                               (cipher)  (messageSize)  (numWrites)  (sslProvider)   Mode  Cnt       Score      Error  Units
SslHandlerEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256            128           10            JDK  thrpt   10  138318,159 ± 4456,938  ops/s
```

This is faster than my 4.1 measurement of `122991.137 ±   974.148`.
With JDK 17 I get `149713,981 ± 8081,987`, while on 4.1 I see `227256,425 ± 12608,357`, so it looks like JDK 17 is helping 4.1 more than 5.

franz1981 commented:
And @chrisvest , the benchmark is supposed to use nio buffers instead unsafe? Just to be sure we are running the same test

chrisvest commented:
> And @chrisvest , the benchmark is supposed to use nio buffers instead unsafe? Just to be sure we are running the same test

That looks like a regression in `main` branch.

chrisvest commented:
> and I've noticed that it uses `ByteBufferMemoryManager` ...is it correct?

That was due to the regression I mentioned. Should be fixed by #12278. That said, I was also using that memory manager, so our numbers are aligned in that sense.

franz1981 commented:
Recycling entries on `PoolThreadCache::allocate`
```java
    private UntetheredMemory allocate(MemoryRegionCache cache, int size) {
        if (cache == null) {
            // no cache found so just return false here
            return null;
        }
        UntetheredMemory allocated = cache.allocate(size, this);
        if (++allocations >= freeSweepAllocationThreshold) {
            allocations = 0;
            trim();
        }
        return allocated;
    }
```
it's using JCTools offering with multiple offered threads ie using a cas operation
that's pretty expensive for what's supposed to be a thread-local allocation cache: we can maybe specialize 
`ObjectPool` (and the inner `Recycler`) to just use a `newChunkedSpscQueue` (instead of newChunked*M*pscQueue), not yet exposed on `PlatformDependent` if the recycling cannot happen from multiple different threads (until a quiescient state is reached -> on the cache disposal its thread local reference `onRemove`).
This should alleviate the cas cost, because the spsc variant is not using any full memory barrier on x86 (and I believe on ARM as well).

![image](https://user-images.githubusercontent.com/13125299/161723224-c0bdfa00-e4d5-4d4f-a9d9-18ddec8f0264.png)

More comments later

franz1981 commented:
I got a different proposal for https://github.com/netty/netty/pull/12221#issuecomment-1088485274, let me send a PR

normanmaurer commented on handler/src/main/java/io/netty5/handler/ssl/EngineWrapper.java:
```diff
@@ -0,0 +1,247 @@
+    private boolean writeBack;
+
+    EngineWrapper(SSLEngine engine, boolean useDirectBuffer) {
+        this.engine = engine;
```
should we add a `Objects.requireNonNull(...)` to fail fast ?

normanmaurer commented on handler/src/main/java/io/netty5/handler/ssl/EngineWrapper.java:
```diff
@@ -0,0 +1,247 @@
+            if (engine instanceof VectoredUnwrap) {
+                VectoredUnwrap vectoredEngine = (VectoredUnwrap) engine;
+                return processResult(vectoredEngine.unwrap(inputs, outputs));
+            } else {
```
nit: you could remove the else

normanmaurer commented on handler/src/main/java/io/netty5/handler/ssl/ReferenceCountedOpenSslContext.java:
```diff
@@ -921,39 +911,46 @@ static long toBIO(ByteBufAllocator allocator, X509Certificate... certChain) thro
+                SSL.freeBIO(bio);
+                throw new IllegalStateException("Could not write data to memory BIO");
+            }
+            return false;
```
shouldn't this be `true` ?

normanmaurer commented on handler/src/main/java/io/netty5/handler/ssl/ReferenceCountedOpenSslEngine.java:
```diff
@@ -517,57 +528,89 @@ private int writePlaintextData(final ByteBuffer src, int len) {
-                sslWrote = SSL.writeToSSL(ssl, memoryAddress(buf), len);
+            try (Buffer buf = alloc.allocate(len)) {
+                buf.writeBytes(src.array(), src.arrayOffset() + pos, len);
+                SslWrite write = new SslWrite(ssl, len);
```
couldn't we re-use the same instance ?

chrisvest replied:
Right, the engine is not accessed concurrently.

normanmaurer commented on handler/src/main/java/io/netty5/handler/ssl/ReferenceCountedOpenSslEngine.java:
```diff
@@ -517,57 +528,89 @@ private int writePlaintextData(final ByteBuffer src, int len) {
+
+        @Override
+        public boolean process(int index, ReadableComponent component) {
+            SSL.bioSetByteBuffer(networkBIO, component.readableNativeAddress(), len, false);
```
I wonder if we need to somehow ensure that the `ReadableComponent` and so its `Buffer` is not GC'ed until we not use the native pointer anymore. Otherwise wee could end up with a crash. Same comment is true for below.

chrisvest replied:
In this case it's up to the caller to ensure. The buffer instance is returned to them.

normanmaurer commented on handler/src/main/java/io/netty5/handler/ssl/ReferenceCountedOpenSslEngine.java:
```diff
@@ -578,22 +621,37 @@ private int readPlaintextData(final ByteBuffer dst) throws SSLException {
-            try {
-                sslRead = SSL.readFromSSL(ssl, memoryAddress(buf), len);
+            try (Buffer buf = alloc.allocate(len)) {
+                SslRead read = new SslRead(ssl, len);
```
same comment as above... cant we use the same instance ?

normanmaurer commented on handler/src/main/java/io/netty5/handler/ssl/ReferenceCountedOpenSslEngine.java:
```diff
@@ -578,22 +621,37 @@ private int readPlaintextData(final ByteBuffer dst) throws SSLException {
+
+        @Override
+        public boolean process(int index, WritableComponent component) {
+            sslRead = SSL.readFromSSL(ssl, component.writableNativeAddress(), len);
```
same comment as above related to GC the component

chrisvest replied:
Like if you call `forEachWritable` and then immediately leak the buffer?
We can make the `forEach` methods include a reachability fence at the end.

chrisvest replied:
I think that should be a separate PR, though.

chrisvest replied:
PR: https://github.com/netty/netty/pull/12302

normanmaurer commented on handler/src/main/java/io/netty5/handler/ssl/SslHandler.java:
```diff
@@ -1197,7 +1118,21 @@ private void handleUnwrapThrowable(ChannelHandlerContext ctx, Throwable cause) {
+                msg = ByteBufBuffer.wrap((ByteBuf) msg);
+            } else {
+                logger.warn("Rejecting buffer: {}", msg);
+                throw new IllegalArgumentException("SslHandler needs Buffer, not ByteBuf: " + msg);
```
Shouldn't we also use `UnsupportedMessageTypeExcetion` here just like above ?

franz1981 commented:
@chrisvest @normanmaurer  just sent https://github.com/netty/netty/pull/12284

chrisvest commented:
> And I suppose the 4.1 results on the same benchmark was using unsafe instead of NIO buffer(s), am I correct?

Just checked. Yes, it has access to Unsafe on both Java 11 and 17. On 17, 4.1 does not have access to the DirectByteBuffer constructor that doesn't take a Cleaner, but that's not stopping 4.1 from using Unsafe.

franz1981 commented:
> but that's not stopping 4.1 from using Unsafe.

That means that is not an Apple 2 Apple comparison here, but still let me know how things change when both run with Unsafe or both without using unsafe buffer(s).
I suppose the most of the cost of the new version will be because of the cleaner usage (masking phantom ref under the hood) now vs nothing before -> for pooled buffers

chrisvest commented:
@franz1981 The two PRs you mentioned in https://github.com/netty/netty/pull/12221#pullrequestreview-930499050 have been brought in via rebase.

@normanmaurer I've addressed your comments, punting 1) removal of implicit buffer conversion, and 2) adding reachability fences to Buffer.forEach, to follow-up PRs.

chrisvest commented:
This is currently generating failures like the following, which needs to be investigated. Possibly being surfaced by the fixes to the `PoolThreadCache` recently.

```
2022-04-06T06:09:02.3512630Z 06:09:02.350 [testsuite-epoll-worker-7-2] WARN  i.n.t.t.socket.SocketSslEchoTest - Unexpected exception from the client side:
2022-04-06T06:09:02.3513514Z io.netty5.handler.codec.DecoderException: java.lang.IllegalArgumentException: newPosition > limit: (777728 > 749384)
2022-04-06T06:09:02.3514448Z 	at io.netty5.handler.codec.ByteToMessageDecoderForBuffer.callDecode(ByteToMessageDecoderForBuffer.java:353)
2022-04-06T06:09:02.3516493Z 	at io.netty5.handler.codec.ByteToMessageDecoderForBuffer.channelInputClosed(ByteToMessageDecoderForBuffer.java:295)
2022-04-06T06:09:02.3517751Z 	at io.netty5.handler.codec.ByteToMessageDecoderForBuffer.channelInputClosed(ByteToMessageDecoderForBuffer.java:268)
2022-04-06T06:09:02.3518815Z 	at io.netty5.handler.codec.ByteToMessageDecoderForBuffer.channelInactive(ByteToMessageDecoderForBuffer.java:251)
2022-04-06T06:09:02.3520207Z 	at io.netty5.handler.ssl.SslHandler.channelInactive(SslHandler.java:919)
2022-04-06T06:09:02.3521046Z 	at io.netty5.channel.DefaultChannelHandlerContext.invokeChannelInactive(DefaultChannelHandlerContext.java:251)
2022-04-06T06:09:02.3522035Z 	at io.netty5.channel.DefaultChannelHandlerContext.findAndInvokeChannelInactive(DefaultChannelHandlerContext.java:246)
2022-04-06T06:09:02.3523147Z 	at io.netty5.channel.DefaultChannelHandlerContext.fireChannelInactive(DefaultChannelHandlerContext.java:233)
2022-04-06T06:09:02.3524000Z 	at io.netty5.channel.ChannelHandler.channelInactive(ChannelHandler.java:229)
2022-04-06T06:09:02.3524859Z 	at io.netty5.channel.DefaultChannelHandlerContext.invokeChannelInactive(DefaultChannelHandlerContext.java:251)
2022-04-06T06:09:02.3525791Z 	at io.netty5.channel.DefaultChannelPipeline.fireChannelInactive(DefaultChannelPipeline.java:806)
2022-04-06T06:09:02.3528542Z 	at io.netty5.channel.AbstractChannel$AbstractUnsafe.lambda$deregister$5(AbstractChannel.java:623)
2022-04-06T06:09:02.3529415Z 	at io.netty5.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:355)
2022-04-06T06:09:02.3530384Z 	at io.netty5.channel.SingleThreadEventLoop.run(SingleThreadEventLoop.java:195)
2022-04-06T06:09:02.3531364Z 	at io.netty5.util.concurrent.SingleThreadEventExecutor.lambda$doStartThread$4(SingleThreadEventExecutor.java:768)
2022-04-06T06:09:02.3532237Z 	at io.netty5.util.internal.ThreadExecutorMap.lambda$apply$1(ThreadExecutorMap.java:68)
2022-04-06T06:09:02.3533029Z 	at io.netty5.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
2022-04-06T06:09:02.3533710Z 	at java.base/java.lang.Thread.run(Thread.java:829)
2022-04-06T06:09:02.3534338Z Caused by: java.lang.IllegalArgumentException: newPosition > limit: (777728 > 749384)
2022-04-06T06:09:02.3535136Z 	at java.base/java.nio.Buffer.createPositionException(Buffer.java:318)
2022-04-06T06:09:02.3535792Z 	at java.base/java.nio.Buffer.position(Buffer.java:293)
2022-04-06T06:09:02.3536421Z 	at java.base/java.nio.ByteBuffer.position(ByteBuffer.java:1094)
2022-04-06T06:09:02.3537119Z 	at io.netty5.buffer.api.internal.Statics.bbsliceFallback(Statics.java:187)
2022-04-06T06:09:02.3537843Z 	at io.netty5.buffer.api.internal.Statics.bbslice(Statics.java:162)
2022-04-06T06:09:02.3538672Z 	at io.netty5.buffer.api.bytebuffer.ByteBufferMemoryManager.sliceMemory(ByteBufferMemoryManager.java:91)
2022-04-06T06:09:02.3539643Z 	at io.netty5.buffer.api.pool.PoolChunk$UntetheredChunkAllocation.memory(PoolChunk.java:586)
2022-04-06T06:09:02.3540440Z 	at io.netty5.buffer.api.pool.PooledBufferAllocator.allocate(PooledBufferAllocator.java:321)
2022-04-06T06:09:02.3541306Z 	at io.netty5.buffer.api.DefaultBufferAllocators$UncloseableBufferAllocator.allocate(DefaultBufferAllocators.java:129)
2022-04-06T06:09:02.3542082Z 	at io.netty5.handler.ssl.SslHandler.allocate(SslHandler.java:2081)
2022-04-06T06:09:02.3542742Z 	at io.netty5.handler.ssl.SslHandler.unwrap(SslHandler.java:1193)
2022-04-06T06:09:02.3543432Z 	at io.netty5.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1080)
2022-04-06T06:09:02.3544118Z 	at io.netty5.handler.ssl.SslHandler.decode(SslHandler.java:1140)
2022-04-06T06:09:02.3545011Z 	at io.netty5.handler.codec.ByteToMessageDecoderForBuffer.decodeRemovalReentryProtection(ByteToMessageDecoderForBuffer.java:379)
2022-04-06T06:09:02.3546035Z 	at io.netty5.handler.codec.ByteToMessageDecoderForBuffer.callDecode(ByteToMessageDecoderForBuffer.java:322)
2022-04-06T06:09:02.3546725Z 	... 17 common frames omitted
```

chrisvest commented:
The allocator issue should be fixed by #12291

franz1981 commented:
@chrisvest still concerned about numbers? 
have you tried the latest change removing the volatile barrier on ArcDrop creation?

chrisvest commented:
@franz1981 I'm personally satisfied with the performance right now. I didn't try the ArcDrop changes.

chrisvest commented:
Two classes of test failures have been looked into here. One only occurs on Windows, and is also not caused by these changes AFAICT. The other _might_ have been fixed by the `duplicate()` call of the last commit, but is also probabilistic, and exists in the code prior to these changes (I believe these changes just make it more likely to occur due to making more use of the new Buffer API), and if it occurs again, we'll have more logging to hopefully narrow it down a bit.
Hence I'm merging this.
